central/group/datastore/datastore_impl.go

```diff
@@ -27,6 +27,57 @@ type dataStoreImpl struct {
 	lock sync.RWMutex
 }
 
+func (ds *dataStoreImpl) Upsert(ctx context.Context, group *storage.Group) error {
+	if ok, err := accessSAC.WriteAllowed(ctx); err != nil {
+		return err
+	} else if !ok {
+		return sac.ErrResourceAccessDenied
+	}
+
+	// Lock to simulate being behind a transaction
+	ds.lock.Lock()
+	defer ds.lock.Unlock()
+
+	if err := ds.validateAndPrepGroupForUpsertNoLock(ctx, group); err != nil {
+		return err
+	}
+
+	return ds.storage.Upsert(ctx, group)
+}
+
+// Validate if the group is allowed to be upserted and prep the group before it is upserted in db.
+// NOTE: This function assumes that the call to this function is already behind a lock.
+func (ds *dataStoreImpl) validateAndPrepGroupForUpsertNoLock(ctx context.Context, newGroup *storage.Group) error {
+	if err := ValidateGroup(newGroup, true); err != nil {
+		return errox.InvalidArgs.CausedBy(err)
+	}
+
+	oldGroup, exists, err := ds.storage.Get(ctx, newGroup.GetProps().GetId())
+	if err != nil {
+		return err
+	}
+	if exists {
+		if err = verifyGroupOriginMatches(ctx, oldGroup); err != nil {
```

Comment:
nit: just reading through this, the name was a bit confusing at first.
Maybe we should get rid of the helper functions and simply have `declarativeconfig.CanModifyResource()` return an `errox.NotAuthorized` with appropriate error message, similar to what we have done with `sac.VerifyAuthZOK`?
This way we have the appropriate HTTP status code the same across datastore and it doesn't require each datastore to redefine the errox type.

Reply:
I think what's good about helper functions is that they allow to:
- specify the type of the resource
- specify the id of the resource
The second one is especially significant since we can't generalize it - the id for the group is accessed via `GetProps.GetId()` while for the rest of resources it's just `GetId`

Regarding helper function names: I agree they are far from ideal :) WDYT about just `verifyGroupOrigin` as an alternative?

Reply:
`verifyGroupOrigin` sounds good to me.

Yeah, I can agree with the type of the resource one, the `ID` one isn't really a reason for me since we could extend the `ResourceWithTraits` interface to include `GetId()`, and since we anyways pass `groups.GetProps()` to the `CanModifyResource()` function, it wouldn't really matter.

But, let's keep the helper for now.